protected/humhub/modules/user/behaviors/Followable.php

```diff
@@ -143,37 +143,47 @@ public function isFollowedByUser($userId = null, $withNotifications = false)
         return false;
     }
 
+    /**
+     * Get a query of users which are followers of this object.
+     *
+     * @param ActiveQueryUser $query e.g. for limit the result
+     * @return ActiveQueryUser
+     */
+    public function getFollowersQuery($query = null)
```

Comment:
Do we need or use the `$query` param anywhere? Limit etc. can be also set on the returned query. I can't find a use case in we'd need to provide a query beforehand. Same with `getFollowingQuery()`.

Reply:
@buddh4 I created the method `getFollowersQuery($query = null)` just to factorize the code from the already existed method `getFollowers($query = null, ...)` which already had the param `$query = null` and used like `$user->getFollowers(User::find()->limit(16))` here https://github.com/humhub/humhub/blob/enh/64-external-users-module/protected/humhub/modules/user/widgets/views/userFollower.php#L6

Reply:
Ok, In this case I'd leave the query param out. The `getFollowers()` does not return a query and therefore its not possible to set a limit otherwise. But in the `getFollowersQuery()` and `getFollowingQuery()` we receive a query and can set such a limit etc. afterwards.

I'd also prefer a similar `public function getFollowers($withNotification = false)` function, but think we can't change this without introducing a breaking change.

Reply:
Can we maybe add the `$withNotifications` in the `getFollowersQuery()` and reuse this in the `getFollowers()` maybe we could even deprecate the old `getFollowers()` in favor of  `getFollowersQuery()`

Reply:
@buddh4 Please check how I implemented this here https://github.com/humhub/humhub/pull/4879/commits/f3019536ea07952abbd4e57a6284cc88f327a6ea.

Reply:
We still should remove the `$query` param here. 

Also we probably should move all the details of this query logic to the `humhub\modules\user\models\Follow` record. The `Follow` class also contains follow related query logic and we should keep all follow related query details in one place. `Followable` should just delegate the calls to the `Follow` model class. Something like `Follow::findFollowersOf($container)` and `Follow::findFollowing($container)`.

Reply:
When further thinking about it I suggest deprecating all old query related functions:

- `Followable::getFollowers()`
- `Followable::getFollowingObjects()`

and not adding any new function to  `Followable`, but instead move (or merge with existing) query function to `Follow` class.

Reply:
Can you also have a look at, not sure if this is a bug:

![grafik](https://user-images.githubusercontent.com/9800854/110077884-34832900-7d87-11eb-984d-816c8e885029.png)

Reply:
> We still should remove the $query param here.

Commit https://github.com/humhub/humhub/pull/4879/commits/5394e00c89fce92a6220d916d5a555a5f127fb8e.

Reply:
@buddh4 

> but instead move or merge with existing query functions in Follow class and use those instead of the deprecated ones.

As I understood you all query methods in the `humhub\modules\user\behaviors\Followable` should be deprecated at all, so instead of current `$this->user->getFollowersQuery()->limit(16)->all()` we will use new `humhub\modules\user\models\Follow::getFollowersQuery($this->user)->limit(16)->all()`.

Note the `Follow::getFollowersQuery(ActiveRecord $target, $withNotifications = null)` already exists with the params, but I see there different query with subquery instead of `LEFT JOIN` as we have in the `Followable->getFollowersQuery()`, so probably we should change this to `LEFT JOIN` as well, or do you think the subquery solution is better?

And probably we should create new `Follow::getFollowingQuery(ActiveRecord $target)` instead of `Followable->getFollowingQuery()` or I should to investigate other methods inside the `Follow`, maybe there some method like `getFollowedContainerIdQuery()` already does same.

Reply:
> As I understood you all query methods in the humhub\modules\user\behaviors\Followable should be deprecated at all, so instead of current $this->user->getFollowersQuery()->limit(16)->all() we will use new humhub\modules\user\models\Follow::getFollowersQuery($this->user)->limit(16)->all().

Yes I'd deprecate the two function and implement the actual logic in `Follow` in order to unify and remove redundance:

- Followable::getFollowers()
- Followable::getFollowingObjects()

Also logic like `getFollowerCount()` should only delegate `Follow` without needing to know about query details.

```
public function getFollowerCount()
    {
        return Follow::find()->where(['object_id' => $this->owner->getPrimaryKey(), 'object_model' => $this->owner->className()])->count();
    }
```

> Followable->getFollowersQuery(), so probably we should change this to LEFT JOIN as well, or do you think the subquery solution is better?

The join is probably faster.

> And probably we should create new Follow::getFollowingQuery(ActiveRecord $target) instead of Followable->getFollowingQuery() or I should to investigate other methods inside the Follow, maybe there some method like getFollowedContainerIdQuery() already does same.

Feel free to make a suggestion what you think is the best way of migrating/moving the logic from `Followable` to  `Follow` if possible without breaking changes and redundence. But if you feel like your prefered solution would require a breaking change, we can also think about it.

Reply:
> Can you also have a look at the following, pretty sure this is a bug, but is probably never called:

I don't find where the method `Followable->beforeDelete()` is used, so I deleted it completely in commit https://github.com/humhub/humhub/pull/4879/commits/b3f1240adabdbb8c3d7325e93f49b9a89bea84a8